File: pkg/tanka/load_test.go
package tanka

import (
	"testing"

	"github.com/grafana/tanka/pkg/kubernetes/manifest"
	"github.com/grafana/tanka/pkg/spec/v1alpha1"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestLoad(t *testing.T) {
	cases := []struct {
		name     string
		baseDir  string
		expected interface{}
		env      *v1alpha1.Environment
	}{
		{
			name:    "static",
			baseDir: "./testdata/cases/withspecjson/",
			expected: manifest.List{{
				"apiVersion": "v1",
				"kind":       "ConfigMap",
				"metadata": map[string]interface{}{
					"name":      "config",
					"namespace": "withspec",
				},
			}},
			env: &v1alpha1.Environment{
				APIVersion: v1alpha1.New().APIVersion,
				Kind:       v1alpha1.New().Kind,
				Metadata: v1alpha1.Metadata{
					Name:      "cases/withspecjson",
					Namespace: "cases/withspecjson/main.jsonnet",
					Labels:    v1alpha1.New().Metadata.Labels,
				},
				Spec: v1alpha1.Spec{
					APIServer: "https://localhost",
					Namespace: "withspec",
				},
				Data: map[string]interface{}{
					"apiVersion": "v1",
					"kind":       "ConfigMap",
					"metadata":   map[string]interface{}{"name": "config", "namespace": "withspec"},
				},
			},
		},
		{
			name:    "static-filename",
			baseDir: "./testdata/cases/withspecjson/main.jsonnet",
			expected: manifest.List{{
				"apiVersion": "v1",
				"kind":       "ConfigMap",
				"metadata": map[string]interface{}{
					"name":      "config",
					"namespace": "withspec",
				},
			}},
			env: &v1alpha1.Environment{
				APIVersion: v1alpha1.New().APIVersion,
				Kind:       v1alpha1.New().Kind,
				Metadata: v1alpha1.Metadata{
					Name:      "cases/withspecjson",
					Namespace: "cases/withspecjson/main.jsonnet",
					Labels:    v1alpha1.New().Metadata.Labels,
				},
				Spec: v1alpha1.Spec{
					APIServer: "https://localhost",
					Namespace: "withspec",
				},
				Data: map[string]interface{}{
					"apiVersion": "v1",
					"kind":       "ConfigMap",
					"metadata":   map[string]interface{}{"name": "config", "namespace": "withspec"},
				},
			},
		},

		{
			name:    "inline",
			baseDir: "./testdata/cases/withenv/",
			expected: manifest.List{{
				"apiVersion": "v1",
				"kind":       "ConfigMap",
				"metadata": map[string]interface{}{
					"name":      "config",
					"namespace": "withenv",
				},
			}},
			env: &v1alpha1.Environment{
				APIVersion: v1alpha1.New().APIVersion,
				Kind:       v1alpha1.New().Kind,
				Metadata: v1alpha1.Metadata{
					Name:      "withenv",
					Namespace: "cases/withenv/main.jsonnet",
					Labels:    v1alpha1.New().Metadata.Labels,
				},
				Spec: v1alpha1.Spec{
					APIServer: "https://localhost",
					Namespace: "withenv",
				},
				Data: map[string]interface{}{
					"apiVersion": "v1",
					"kind":       "ConfigMap",
					"metadata":   map[string]interface{}{"name": "config", "namespace": "withenv"},
				},
			},
		},
		{
			name:    "inline-filename",
			baseDir: "./testdata/cases/withenv/main.jsonnet",
			expected: manifest.List{{
				"apiVersion": "v1",
				"kind":       "ConfigMap",
				"metadata": map[string]interface{}{
					"name":      "config",
					"namespace": "withenv",
				},
			}},
			env: &v1alpha1.Environment{
				APIVersion: v1alpha1.New().APIVersion,
				Kind:       v1alpha1.New().Kind,
				Metadata: v1alpha1.Metadata{
					Name:      "withenv",
					Namespace: "cases/withenv/main.jsonnet",
					Labels:    v1alpha1.New().Metadata.Labels,
				},
				Spec: v1alpha1.Spec{
					APIServer: "https://localhost",
					Namespace: "withenv",
				},
				Data: map[string]interface{}{
					"apiVersion": "v1",
					"kind":       "ConfigMap",
					"metadata":   map[string]interface{}{"name": "config", "namespace": "withenv"},
				},
			},
		},
	}

	for _, test := range cases {
		t.Run(test.name, func(t *testing.T) {
			l, err := Load(test.baseDir, Opts{})
			require.NoError(t, err)

			assert.Equal(t, test.expected, l.Resources)
			assert.Equal(t, test.env, l.Env)
		})
	}
}

func TestLoadSelectEnvironment(t *testing.T) {
	// No match
	_, err := Load("./testdata/cases/multiple-inline-envs", Opts{Name: "no match"})
	assert.EqualError(t, err, "found no matching environments; run 'tk env list ./testdata/cases/multiple-inline-envs' to view available options")

	// Empty options, match all environments
	_, err = Load("./testdata/cases/multiple-inline-envs", Opts{})
	assert.EqualError(t, err, "found multiple Environments in './testdata/cases/multiple-inline-envs'. Use `--name` to select a single one: \n - project1-env1\n - project1-env2\n - project2-env1")

	// Partial match two environments
	_, err = Load("./testdata/cases/multiple-inline-envs", Opts{Name: "env1"})
	assert.EqualError(t, err, "found multiple Environments in './testdata/cases/multiple-inline-envs'. Use `--name` to select a single one: \n - project1-env1\n - project2-env1")

	// Partial match
	result, err := Load("./testdata/cases/multiple-inline-envs", Opts{Name: "project2"})
	assert.NoError(t, err)
	assert.Equal(t, "project2-env1", result.Env.Metadata.Name)

	// Full match
	result, err = Load("./testdata/cases/multiple-inline-envs", Opts{Name: "project1-env1"})
	assert.NoError(t, err)
	assert.Equal(t, "project1-env1", result.Env.Metadata.Name)
}

func TestLoadSelectEnvironmentFullMatchHasPriority(t *testing.T) {
	// `base` matches both `base` and `base-and-more`
	// However, the full match should win
	result, err := Load("./testdata/cases/inline-name-conflict", Opts{Name: "base"})
	assert.NoError(t, err)
	assert.Equal(t, "base", result.Env.Metadata.Name)
}

func TestLoadFailsWhenBothSpecAndInline(t *testing.T) {
	_, err := Load("./testdata/cases/static-and-inline", Opts{Name: "inline"})
	assert.EqualError(t, err, "found a tanka Environment resource. Check that you aren't using a spec.json and inline environments simultaneously")
}

File: rust-export/src/load_test.rs
use crate::loader::{detect_loader, LoaderOpts};
use serde_json::json;
use std::path::PathBuf;

#[test]
fn test_load_static() {
    let base_dir = PathBuf::from("testdata/cases/withspecjson");
    let loader = detect_loader(&base_dir, None).expect("Failed to detect loader");
    assert_eq!(loader.name(), "static");

    let loaded = loader
        .load(&base_dir, &LoaderOpts::default())
        .expect("Failed to load environment");

    // Check environment metadata
    assert_eq!(
        loaded.environment.api_version,
        "tanka.dev/v1alpha1",
        "API version mismatch"
    );
    assert_eq!(loaded.environment.kind, "Environment", "Kind mismatch");
    assert_eq!(
        loaded.environment.metadata.name,
        Some("withspec".to_string()),
        "Name mismatch"
    );
    assert_eq!(
        loaded.environment.spec.api_server,
        Some("https://localhost".to_string()),
        "API server mismatch"
    );
    assert_eq!(loaded.environment.spec.namespace, "withspec");

    // Check data (manifests)
    assert!(loaded.environment.data.is_some(), "Data is missing");
    let data = loaded.environment.data.unwrap();

    let expected_manifest = json!({
        "apiVersion": "v1",
        "kind": "ConfigMap",
        "metadata": {
            "name": "config"
        }
    });

    assert_eq!(data, expected_manifest, "Manifest mismatch");
}

#[test]
fn test_load_static_filename() {
    let base_dir = PathBuf::from("testdata/cases/withspecjson/main.jsonnet");
    let loader = detect_loader(&base_dir, None).expect("Failed to detect loader");
    assert_eq!(loader.name(), "static");

    let loaded = loader
        .load(&base_dir.parent().unwrap(), &LoaderOpts::default())
        .expect("Failed to load environment");

    // Check environment metadata
    assert_eq!(
        loaded.environment.api_version,
        "tanka.dev/v1alpha1",
        "API version mismatch"
    );
    assert_eq!(loaded.environment.kind, "Environment", "Kind mismatch");
    assert_eq!(
        loaded.environment.metadata.name,
        Some("withspec".to_string()),
        "Name mismatch"
    );

    // Check data
    assert!(loaded.environment.data.is_some(), "Data is missing");
}

#[test]
fn test_load_inline() {
    let base_dir = PathBuf::from("testdata/cases/withenv");
    let loader = detect_loader(&base_dir, None).expect("Failed to detect loader");
    assert_eq!(loader.name(), "inline");

    let loaded = loader
        .load(&base_dir, &LoaderOpts::default())
        .expect("Failed to load environment");

    // Check environment metadata
    assert_eq!(
        loaded.environment.api_version,
        "tanka.dev/v1alpha1",
        "API version mismatch"
    );
    assert_eq!(loaded.environment.kind, "Environment", "Kind mismatch");
    assert_eq!(
        loaded.environment.metadata.name,
        Some("withenv".to_string()),
        "Name mismatch"
    );
    assert_eq!(
        loaded.environment.spec.api_server,
        Some("https://localhost".to_string()),
        "API server mismatch"
    );
    assert_eq!(loaded.environment.spec.namespace, "withenv");

    // Check data (manifests)
    assert!(loaded.environment.data.is_some(), "Data is missing");
    let data = loaded.environment.data.unwrap();

    let expected_manifest = json!({
        "apiVersion": "v1",
        "kind": "ConfigMap",
        "metadata": {
            "name": "config"
        }
    });

    assert_eq!(data, expected_manifest, "Manifest mismatch");
}

#[test]
fn test_load_inline_filename() {
    let base_dir = PathBuf::from("testdata/cases/withenv/main.jsonnet");
    let loader = detect_loader(&base_dir, None).expect("Failed to detect loader");
    assert_eq!(loader.name(), "inline");

    let loaded = loader
        .load(&base_dir, &LoaderOpts::default())
        .expect("Failed to load environment");

    // Check environment metadata
    assert_eq!(
        loaded.environment.api_version,
        "tanka.dev/v1alpha1",
        "API version mismatch"
    );
    assert_eq!(loaded.environment.kind, "Environment", "Kind mismatch");
    assert_eq!(
        loaded.environment.metadata.name,
        Some("withenv".to_string()),
        "Name mismatch"
    );

    // Check data
    assert!(loaded.environment.data.is_some(), "Data is missing");
}

#[test]
fn test_peek_static() {
    let base_dir = PathBuf::from("testdata/cases/withspecjson");
    let loader = detect_loader(&base_dir, None).expect("Failed to detect loader");

    let env = loader
        .peek(&base_dir, &LoaderOpts::default())
        .expect("Failed to peek environment");

    // Peek should not load data
    assert!(env.data.is_none(), "Peek should not load data");
    assert_eq!(env.metadata.name, Some("withspec".to_string()));
}

#[test]
fn test_peek_inline() {
    let base_dir = PathBuf::from("testdata/cases/withenv");
    let loader = detect_loader(&base_dir, None).expect("Failed to detect loader");

    let env = loader
        .peek(&base_dir, &LoaderOpts::default())
        .expect("Failed to peek environment");

    // Peek should not load data
    assert!(env.data.is_none(), "Peek should not load data");
    assert_eq!(env.metadata.name, Some("withenv".to_string()));
}

#[test]
fn test_list_static() {
    let base_dir = PathBuf::from("testdata/cases/withspecjson");
    let loader = detect_loader(&base_dir, None).expect("Failed to detect loader");

    let envs = loader
        .list(&base_dir, &LoaderOpts::default())
        .expect("Failed to list environments");

    assert_eq!(envs.len(), 1, "Static should return exactly one environment");
    assert_eq!(envs[0].metadata.name, Some("withspec".to_string()));
}

#[test]
fn test_list_inline() {
    let base_dir = PathBuf::from("testdata/cases/withenv");
    let loader = detect_loader(&base_dir, None).expect("Failed to detect loader");

    let envs = loader
        .list(&base_dir, &LoaderOpts::default())
        .expect("Failed to list environments");

    assert_eq!(
        envs.len(),
        1,
        "Inline should return at least one environment"
    );
    assert_eq!(envs[0].metadata.name, Some("withenv".to_string()));
}

#[test]
fn test_eval_static() {
    let base_dir = PathBuf::from("testdata/cases/withspecjson");
    let loader = detect_loader(&base_dir, None).expect("Failed to detect loader");

    let data = loader
        .eval(&base_dir, &LoaderOpts::default())
        .expect("Failed to eval");

    // Should return raw jsonnet output (the ConfigMap)
    let expected = json!({
        "apiVersion": "v1",
        "kind": "ConfigMap",
        "metadata": {
            "name": "config"
        }
    });

    assert_eq!(data, expected);
}

#[test]
fn test_eval_inline() {
    let base_dir = PathBuf::from("testdata/cases/withenv");
    let loader = detect_loader(&base_dir, None).expect("Failed to detect loader");

    let data = loader
        .eval(&base_dir, &LoaderOpts::default())
        .expect("Failed to eval");

    // Should return the full Environment object
    assert!(data.is_object());
    let obj = data.as_object().unwrap();
    assert_eq!(obj.get("kind").unwrap(), "Environment");
}

